breadbox/breadbox/crud/dataset.py

```diff
@@ -794,19 +816,60 @@ def get_unique_dimension_ids_from_datasets(
         .all()
     }
     # Get all tabular identifiers for that dimension type
-    tabular_dimension_ids = {
+    tabular_given_ids = {
         given_id for (given_id,) in
-        db.query(distinct(TabularCell.dimension_given_id))
-        .join(TabularColumn)
+        db.query(TabularColumn)
         .filter(
             and_(
                 TabularColumn.dataset_id.in_(dataset_ids),
                 TabularColumn.given_id == dimension_type.id_column,
                 TabularColumn.dataset_dimension_type == dimension_type.name,
             )
         )
+        .join(TabularCell)
+        .distinct(TabularCell.dimension_given_id)
+        .with_entities(TabularCell.dimension_given_id)
         .all()
     }
     # Combine the unique given ids from the tabular and matrix datasets
-    return tabular_dimension_ids.union(matrix_dimension_ids)
+    return matrix_given_ids.union(tabular_given_ids)
+
 
+def get_metadata_used_in_matrix_dataset(
+    db: SessionWithUser, 
+    dimension_type: DimensionType,
+    matrix_dataset: MatrixDataset, 
+    dimension_subtype_cls: Union[Type[DatasetFeature], Type[DatasetSample]],
+    metadata_col_name: str
+) -> dict[str, str]:
+    """
+    For the given matrix dataset, load a column from the associated metadata.
+    The result will only include given ids which exist in both the dataset and the metadata.
+    For example, if a dataset's sample type is "depmap_model", and the requested metadata field name is "label",
+    then this will return a dictionary with depmap ids as keys and cell line names as values.
+    """
+    assert_user_has_access_to_dataset(matrix_dataset, db.user)
+    matrix_dataset_id = matrix_dataset.id
+
+    # Using a subquery makes this MUCH faster than two separate queries would be 
+    # because it reduces the number of rows that need to be fetched and constructed 
+    # into python objects (which is usually by far the slowest part of SQLAlchemy queries).
+    given_id_subquery = (
+        db.query(dimension_subtype_cls.given_id)
+        .filter(dimension_subtype_cls.dataset_id == matrix_dataset_id)
+        .subquery()
+    )
+    metadata_vals_by_id = (
+        db.query(TabularColumn)
+        .filter(
+            and_(
+                TabularColumn.dataset_id == dimension_type.dataset_id,
+                TabularColumn.given_id == metadata_col_name,
+            )
+        )
+        .join(TabularCell)
+        .filter(TabularCell.dimension_given_id.in_(given_id_subquery))
```

Comment:
I don't feel 100% about this pattern of filtering by `._in(subquery)` - which has the potential to be slow with very large numbers of IDs. However, it's easy to read, and this change made this particular operation more than 5x faster - it's now taking a negligable amount of time (0.08 seconds) for even with 2-3k given IDs.

If we do this operation on any datasets that are an order of magnitude larger, it might be worth revisiting.

Reply:
Why do you think `.in(subquery)` is slow?

Reply:
You could also write this with an `exists` clause instead of `in` clause would give you similar readability but I expect the performance to be comparable. (Modulo the quality of the query optimizer)

Reply:
Actually yeah, thinking a bit more about this, it's really not an issue in this case. I have a placeholder in my mind that subqueries are slow and/or bad practice. In reality, it's really just correlated subqueries that are slow - where you use a value from the parent query, which forces the subquery to be re-executed every time its used (not the case here)

I tend to think of `IN` as being slow too - but you're right, that's not necessarily true, and using `EXISTS` would address that problem.